#### <h1 style="color: #ab9bde;">IMPORT LIBRARIES</h1>

In [ ]:
# Core Libraries
import numpy as np
import pandas as pd

# Visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Plot settings
plt.style.use('default')
sns.set_theme(style="whitegrid")

# Statistics
from scipy import stats

# Date & Time
from datetime import datetime

# Machine Learning Utilities
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder, RobustScaler
from sklearn.pipeline import Pipeline


# Display Settings
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)
pd.set_option('display.float_format', '{:.2f}'.format)

# Warning Control
import warnings
warnings.filterwarnings('ignore')

print("All libraries imported successfully.")

All libraries imported successfully.


<div style="color:#ab9bde;">

## <h2>Load EDA enriched dataset:</h2>
</div>

In [3]:
df = pd.read_csv("/Users/laxmigupte/Desktop/Hotel-Revenue-Operations/data/hotel_bookings_eda.csv")
print("EDA enriched Dataset loaded successfully.")

EDA enriched Dataset loaded successfully.


In [4]:
df.head()

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,meal,country,market_segment,distribution_channel,is_repeated_guest,previous_cancellations,previous_bookings_not_canceled,reserved_room_type,assigned_room_type,booking_changes,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,lead_time_bucket,adr_bucket,total_nights,estimated_booking_value
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,0.00,0,BB,PRT,Direct,Direct,0,0,0,C,C,3,No Deposit,NaN,NaN,0,Transient,0.00,0,0,Check-Out,2015-07-01,181-365 Days,0-50,0,0.00
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,0.00,0,BB,PRT,Direct,Direct,0,0,0,C,C,4,No Deposit,NaN,NaN,0,Transient,0.00,0,0,Check-Out,2015-07-01,365+ Days,0-50,0,0.00
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,0.00,0,BB,GBR,Direct,Direct,0,0,0,A,C,0,No Deposit,NaN,NaN,0,Transient,75.00,0,0,Check-Out,2015-07-02,0-30 Days,51-100,1,75.00
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,0.00,0,BB,GBR,Corporate,Corporate,0,0,0,A,A,0,No Deposit,304.00,NaN,0,Transient,75.00,0,0,Check-Out,2015-07-02,0-30 Days,51-100,1,75.00
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,0.00,0,BB,GBR,Online TA,TA/TO,0,0,0,A,A,0,No Deposit,240.00,NaN,0,Transient,98.00,0,1,Check-Out,2015-07-03,0-30 Days,51-100,2,196.00


In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 87396 entries, 0 to 87395
Data columns (total 36 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   hotel                           87396 non-null  str    
 1   is_canceled                     87396 non-null  int64  
 2   lead_time                       87396 non-null  int64  
 3   arrival_date_year               87396 non-null  int64  
 4   arrival_date_month              87396 non-null  str    
 5   arrival_date_week_number        87396 non-null  int64  
 6   arrival_date_day_of_month       87396 non-null  int64  
 7   stays_in_weekend_nights         87396 non-null  int64  
 8   stays_in_week_nights            87396 non-null  int64  
 9   adults                          87396 non-null  int64  
 10  children                        87392 non-null  float64
 11  babies                          87396 non-null  int64  
 12  meal                            87396 non-n

In [6]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
is_canceled,87396.00,0.27,0.45,0.00,0.00,0.00,1.00,1.00
lead_time,87396.00,79.89,86.05,0.00,11.00,49.00,125.00,737.00
arrival_date_year,87396.00,2016.21,0.69,2015.00,2016.00,2016.00,2017.00,2017.00
arrival_date_week_number,87396.00,26.84,13.67,1.00,16.00,27.00,37.00,53.00
arrival_date_day_of_month,87396.00,15.82,8.84,1.00,8.00,16.00,23.00,31.00
stays_in_weekend_nights,87396.00,1.01,1.03,0.00,0.00,1.00,2.00,19.00
stays_in_week_nights,87396.00,2.63,2.05,0.00,1.00,2.00,4.00,50.00
adults,87396.00,1.88,0.63,0.00,2.00,2.00,2.00,55.00
children,87392.00,0.14,0.46,0.00,0.00,0.00,0.00,10.00
babies,87396.00,0.01,0.11,0.00,0.00,0.00,0.00,10.00


In [7]:
df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date', 'lead_time_bucket',
       'adr_bucket', 'total_nights', 'estimated_booking_value'],
      dtype='str')

In [8]:
required_columns = [
    "is_canceled",
    "lead_time",
    "adr",
    "stays_in_weekend_nights",
    "stays_in_week_nights",
    "adults",
    "children",
    "babies",
    "previous_cancellations",
    "booking_changes",
    "total_of_special_requests",
    "deposit_type",
    "customer_type",
    "market_segment",
    "hotel",
    "arrival_date_month"
]

missing_required_columns = [
    col for col in required_columns if col not in df.columns
]

missing_required_columns

if len(missing_required_columns) == 0:
    print("All required columns are available.")
else:
    print("Missing columns:")
    print(missing_required_columns)

All required columns are available.


<div style="color:#ab9bde;">

## <h2>Feature Engineering Plan:</h2>

This notebook builds predictive signals for hotel booking cancellation modeling.

The goal is to transform raw booking data into meaningful features that can help identify cancellation risk and support a future hotel revenue assistant.

This notebook follows the structure:

1. Temporal features
2. Guest behavioural signals
3. Revenue proxy features
4. Encoding and scaling preparation
5. Sklearn pipeline architecture

</div>

#### <h3 style="color: #ab9bde;">Validate Existing Engineered Features</h3>

In [9]:
existing_engineered_features = [
    "total_nights",
    "estimated_booking_value",
    "lead_time_bucket",
    "adr_bucket"
]

for col in existing_engineered_features:
    print(f"{col}: {col in df.columns}")

total_nights: True
estimated_booking_value: True
lead_time_bucket: True
adr_bucket: True


In [10]:
# Validate total nights
total_nights_check = (
    df["total_nights"] ==
    df["stays_in_weekend_nights"] + df["stays_in_week_nights"]
)

print(f"Correct total_nights rows: {total_nights_check.sum():,}")
print(f"Incorrect total_nights rows: {(~total_nights_check).sum():,}")

Correct total_nights rows: 87,396
Incorrect total_nights rows: 0


In [11]:
# Validate estimated booking value
estimated_value_check = np.isclose(
    df["estimated_booking_value"],
    df["adr"] * df["total_nights"]
)

print(f"Correct estimated_booking_value rows: {estimated_value_check.sum():,}")
print(f"Incorrect estimated_booking_value rows: {(~estimated_value_check).sum():,}")

Correct estimated_booking_value rows: 87,396
Incorrect estimated_booking_value rows: 0


In [12]:
# Missing values
df[existing_engineered_features].isna().sum()

total_nights               0
estimated_booking_value    0
lead_time_bucket           0
adr_bucket                 1
dtype: int64

In [13]:
# Bucket distribution
df["lead_time_bucket"].value_counts(dropna=False)

lead_time_bucket
0-30 Days       34644
31-90 Days      22744
91-180 Days     18243
181-365 Days    11200
365+ Days         565
Name: count, dtype: int64

In [14]:
df["adr_bucket"].value_counts(dropna=False)

adr_bucket
51-100     35793
101-150    26632
151-200    10395
0-50        9857
201-300     4457
300+         261
NaN            1
Name: count, dtype: int64

In [15]:
df[df["adr_bucket"].isna()][
    [
        "adr",
        "adr_bucket",
        "total_nights",
        "estimated_booking_value",
        "is_canceled",
        "hotel",
        "market_segment"
    ]
]

,adr,adr_bucket,total_nights,estimated_booking_value,is_canceled,hotel,market_segment
11591,-6.38,NaN,10,-63.80,0,Resort Hotel,Groups


In [16]:
df["has_invalid_adr"] = np.where(df["adr"] < 0, 1, 0)

df["has_invalid_adr"].value_counts()

has_invalid_adr
0    87395
1        1
Name: count, dtype: int64

In [17]:
# Convert adr_bucket to object type before filling missing bucket
df["adr_bucket"] = df["adr_bucket"].astype("object")

# Fill the missing bucket caused by invalid negative ADR
df["adr_bucket"] = df["adr_bucket"].fillna("Invalid ADR")

In [18]:
df["adr_bucket"].value_counts(dropna=False)

adr_bucket
51-100         35793
101-150        26632
151-200        10395
0-50            9857
201-300         4457
300+             261
Invalid ADR        1
Name: count, dtype: int64

In [19]:
df["adr_bucket"].isna().sum()

np.int64(0)

<div style="color:#ab9bde;">

## <h2>Insights:</h2>

- The `adr_bucket` column had one missing value because one booking had a negative ADR value of `-6.38`. Since the ADR bucket ranges start from zero, this row did not fall into any existing bucket.

- Instead of dropping the row at this stage, the row was flagged using `has_invalid_adr`, and the missing ADR bucket was labeled as `Invalid ADR`.

- This keeps the dataset complete while making the data issue explicit for future modeling decisions. </div>

#### <h3 style="color: #ab9bde;">Stay Duration Features</h3>

In [20]:
# Create zero-night stay flag

df["is_zero_night_stay"] = np.where(
    df["total_nights"] == 0,
    1,
    0
)

df["is_zero_night_stay"].value_counts()

is_zero_night_stay
0    86745
1      651
Name: count, dtype: int64

In [21]:
# Create weekend stay flag

df["is_weekend_stay"] = np.where(
    df["stays_in_weekend_nights"] > 0,
    1,
    0
)

df["is_weekend_stay"].value_counts()

is_weekend_stay
1    52245
0    35151
Name: count, dtype: int64

In [22]:
# Create stay length category

def categorize_stay_length(nights):
    if nights == 0:
        return "Zero-night stay"
    elif nights <= 2:
        return "Short stay"
    elif nights <= 5:
        return "Medium stay"
    elif nights <= 10:
        return "Long stay"
    else:
        return "Extended stay"

df["stay_length_category"] = df["total_nights"].apply(categorize_stay_length)

df["stay_length_category"].value_counts()

stay_length_category
Medium stay        38156
Short stay         32929
Long stay          13721
Extended stay       1939
Zero-night stay      651
Name: count, dtype: int64

In [23]:
# Validate
stay_duration_features = [
    "total_nights",
    "is_zero_night_stay",
    "is_weekend_stay",
    "stay_length_category"
]

df[stay_duration_features].isna().sum()

total_nights            0
is_zero_night_stay      0
is_weekend_stay         0
stay_length_category    0
dtype: int64

<div style="color:#ab9bde;">

## <h2>Insights:</h2>

Three booking-structure features were created:

- `is_zero_night_stay`
- `is_weekend_stay`
- `stay_length_category`

These features convert raw stay-night columns into clearer booking signals that can be used later during preprocessing and modeling.
</div>

#### <h3 style="color: #ab9bde;">Temporal Features</h3>

In [24]:
# Create

df["is_last_minute_booking"] = np.where(df["lead_time"] <= 7, 1, 0)
df["is_long_lead_booking"] = np.where(df["lead_time"] >= 90, 1, 0)

In [25]:
# Numberise the months
month_mapping = {
    "January": 1,
    "February": 2,
    "March": 3,
    "April": 4,
    "May": 5,
    "June": 6,
    "July": 7,
    "August": 8,
    "September": 9,
    "October": 10,
    "November": 11,
    "December": 12
}

df["arrival_month_number"] = df["arrival_date_month"].map(month_mapping)

In [26]:
# Season
def assign_season(month):
    if month in ["December", "January", "February"]:
        return "Winter"
    elif month in ["March", "April", "May"]:
        return "Spring"
    elif month in ["June", "July", "August"]:
        return "Summer"
    else:
        return "Autumn"

df["arrival_season"] = df["arrival_date_month"].apply(assign_season)

In [27]:
# Peak season
peak_months = ["April", "May", "June", "July", "August"]

df["is_peak_season"] = np.where(
    df["arrival_date_month"].isin(peak_months),
    1,
    0
)

#### <h3 style="color: #ab9bde;">Guest Behavioural Signals</h3>

In [28]:
# Create behavioural features
df["has_previous_cancellation"] = np.where(
    df["previous_cancellations"] > 0,
    1,
    0
)

df["has_previous_successful_booking"] = np.where(
    df["previous_bookings_not_canceled"] > 0,
    1,
    0
)

df["has_booking_changes"] = np.where(
    df["booking_changes"] > 0,
    1,
    0
)

df["has_special_requests"] = np.where(
    df["total_of_special_requests"] > 0,
    1,
    0
)

df["is_waitlisted"] = np.where(
    df["days_in_waiting_list"] > 0,
    1,
    0
)

In [29]:
# Previous cancellation rate
df["previous_total_bookings"] = (
    df["previous_cancellations"] + 
    df["previous_bookings_not_canceled"]
)

df["previous_cancellation_rate"] = np.where(
    df["previous_total_bookings"] > 0,
    df["previous_cancellations"] / df["previous_total_bookings"],
    0
)

In [30]:
# Interaction with deposit
df["repeat_guest_with_deposit"] = np.where(
    (df["is_repeated_guest"] == 1) & (df["deposit_type"] != "No Deposit"),
    1,
    0
)

df["repeat_guest_non_refund"] = np.where(
    (df["is_repeated_guest"] == 1) & (df["deposit_type"] == "Non Refund"),
    1,
    0
)

#### <h3 style="color: #ab9bde;">Revenue Proxy Features</h3>

In [31]:
df["is_zero_adr"] = np.where(df["adr"] == 0, 1, 0)

df["is_high_adr"] = np.where(
    df["adr"] >= df["adr"].quantile(0.75),
    1,
    0
)

df["is_high_value_booking"] = np.where(
    df["estimated_booking_value"] >= df["estimated_booking_value"].quantile(0.75),
    1,
    0
)

In [32]:
df["room_type_changed"] = np.where(
    df["reserved_room_type"] != df["assigned_room_type"],
    1,
    0
)

In [33]:
df["reserved_room_type"].value_counts()

reserved_room_type
A    56552
D    17398
E     6049
F     2823
G     2052
B      999
C      915
H      596
L        6
P        6
Name: count, dtype: int64

 <div style="color:#ab9bde;">

## <h2>Room Type Premium Flag Decision:</h2>
 

- The dataset contains room type codes such as `A`, `B`, `C`, `D`, `E`, `F`, `G`, `H`, `L`, and `P`.

- However, the dataset does not provide a clear mapping that explains which room types are standard, premium, or luxury. Because of this, a `room_type_premium_flag` is not created at this stage.

- Instead, the feature `room_type_changed` is kept because it is directly observable and does not require assumptions. </div>

#### <h3 style="color: #ab9bde;">Leakage Review</h3>

In [37]:
target_col = "is_canceled"

leakage_cols = [
    "reservation_status",
    "reservation_status_date"
]

#### <h3 style="color: #ab9bde;">Feature Selection Setup</h3>

In [34]:
candidate_features = [
    # original useful features
    "hotel",
    "lead_time",
    "market_segment",
    "distribution_channel",
    "deposit_type",
    "customer_type",
    "adr",
    "total_of_special_requests",
    "booking_changes",
    "previous_cancellations",
    "previous_bookings_not_canceled",
    "is_repeated_guest",
    "required_car_parking_spaces",
    "total_nights",
    
    # engineered booking structure
    "stay_length_category",
    "is_zero_night_stay",
    "is_weekend_stay",

    # engineered temporal
    "lead_time_bucket",
    "is_last_minute_booking",
    "is_long_lead_booking",
    "arrival_month_number",
    "arrival_season",
    "is_peak_season",

    # engineered behavioural
    "has_previous_cancellation",
    "has_previous_successful_booking",
    "has_booking_changes",
    "has_special_requests",
    "is_waitlisted",
    "previous_total_bookings",
    "previous_cancellation_rate",
    "repeat_guest_with_deposit",
    "repeat_guest_non_refund",

    # engineered revenue
    "estimated_booking_value",
    "adr_bucket",
    "has_invalid_adr",
    "is_zero_adr",
    "is_high_adr",
    "is_high_value_booking",
    "room_type_changed"
]

In [38]:
X = df[candidate_features].copy()
y = df[target_col].copy()

print(X.shape)
print(y.shape)

(87396, 39)
(87396,)


 <div style="color:#ab9bde;">

## <h2>Feature Selection Setup Check:</h2>
 

The candidate feature matrix now includes the engineered stay-duration features:

- `stay_length_category`
- `is_zero_night_stay`
- `is_weekend_stay`

The feature matrix `X` and target vector `y` have matching row counts, confirming that the selected features are aligned with the target variable. </div>

#### <h3 style="color: #ab9bde;">Encode & Scale</h3>

In [39]:
# Numerical features: continuous or count-based values

numeric_features = [
    "lead_time",
    "adr",
    "total_nights",
    "estimated_booking_value",
    "total_of_special_requests",
    "booking_changes",
    "previous_cancellations",
    "previous_bookings_not_canceled",
    "previous_total_bookings",
    "previous_cancellation_rate",
    "required_car_parking_spaces",
    "arrival_month_number"
]

In [40]:
# Nominal categorical features: categories with no natural order

nominal_features = [
    "hotel",
    "market_segment",
    "distribution_channel",
    "deposit_type",
    "customer_type",
    "arrival_season"
]

In [41]:
# Ordinal categorical features: categories with meaningful order

ordinal_features = [
    "lead_time_bucket",
    "adr_bucket",
    "stay_length_category"
]

In [42]:
# Binary features: already encoded as 0/1

binary_features = [
    "is_repeated_guest",
    "is_last_minute_booking",
    "is_long_lead_booking",
    "is_peak_season",
    "has_previous_cancellation",
    "has_previous_successful_booking",
    "has_booking_changes",
    "has_special_requests",
    "is_waitlisted",
    "repeat_guest_with_deposit",
    "repeat_guest_non_refund",
    "has_invalid_adr",
    "is_zero_adr",
    "is_high_adr",
    "is_high_value_booking",
    "room_type_changed",
    "is_zero_night_stay",
    "is_weekend_stay"
]

In [43]:
# Validate the existence of grouped features
all_preprocessing_features = (
    numeric_features +
    nominal_features +
    ordinal_features +
    binary_features
)

missing_features = [
    col for col in all_preprocessing_features
    if col not in X.columns
]

missing_features

[]

In [44]:
# Check missed candidate features
unused_candidate_features = [
    col for col in candidate_features
    if col not in all_preprocessing_features
]

unused_candidate_features

[]

#### <h3 style="color: #ab9bde;">Sklearn Pipeline Architecture</h3>

In [46]:
# Define ordinal category order
lead_time_order = [
    "0–30 Days",
    "31–90 Days",
    "91–180 Days",
    "181–365 Days",
    "365+ Days"
]

adr_order = [
    "0–50",
    "51–100",
    "101–150",
    "151–200",
    "201–300",
    "300+",
    "Invalid ADR"
]

stay_length_order = [
    "Zero-night stay",
    "Short stay",
    "Medium stay",
    "Long stay",
    "Extended stay"
]

ordinal_categories = [
    lead_time_order,
    adr_order,
    stay_length_order
]

In [47]:
# Build the preprocessor
preprocessor = ColumnTransformer(
    transformers=[
        (
            "numeric",
            RobustScaler(),
            numeric_features
        ),
        (
            "nominal",
            OneHotEncoder(handle_unknown="ignore"),
            nominal_features
        ),
        (
            "ordinal",
            OrdinalEncoder(
                categories=ordinal_categories,
                handle_unknown="use_encoded_value",
                unknown_value=-1
            ),
            ordinal_features
        ),
        (
            "binary",
            "passthrough",
            binary_features
        )
    ],
    remainder="drop"
)

In [48]:
# Architecture
X_preprocessed = preprocessor.fit_transform(X)

print(f"Original feature matrix shape: {X.shape}")
print(f"Preprocessed feature matrix shape: {X_preprocessed.shape}")

Original feature matrix shape: (87396, 39)
Preprocessed feature matrix shape: (87396, 59)


 

 <div style="color:#ab9bde;">

## <h2>Sklearn Pipeline Architecture Check:</h2>

- The preprocessing architecture was successfully tested.

- The original feature matrix contained 87,396 rows and 39 candidate features.

- After preprocessing, the matrix contains 87,396 rows and 59 transformed features. The increase in feature count is expected because nominal categorical features were expanded using `OneHotEncoder`.

- This confirms that the preprocessing setup can handle numerical, nominal, ordinal, and binary features correctly. </div>



### <h3 style="color: #ab9bde;">Featured Engineered Dataset</h3>

In [49]:
OUTPUT_PATH = "../data/hotel_bookings_feature_engineered.csv"

df.to_csv(OUTPUT_PATH, index=False)

print(f"Feature-engineered dataset saved successfully to: {OUTPUT_PATH}")
print(f"Rows saved: {df.shape[0]:,}")
print(f"Columns saved: {df.shape[1]:,}")

Feature-engineered dataset saved successfully to: ../data/hotel_bookings_feature_engineered.csv
Rows saved: 87,396
Columns saved: 58
